### ISSUE TITLE
Create/destroy subvolumes for chroots on btrfs storage instead of using directories

### ISSUE THREAD

**Issue by Conan-Kudo** [rpm-software-management/mock#1146]
### Short description of the problem

Fedora infrastructure has encountered the problem that 32-bit builds are failing due to inode counts exceeding the limit for 32-bit numbers. The Fedora Koji builders are on Btrfs, and the upstream recommendation is that each chroot should be in their own subvolume to take advantage of the property that inode counts are independent per subvolume.

### Output of `rpm -q mock`

`mock-4.1-1.fc38.noarch`

### Steps to reproduce issue

This is not easy to reproduce locally, it's basically: keep building stuff until the inode count goes over the 32-bit limit.

### Any additional notes

More details of the problem:

* Fedora releng ticket: https://pagure.io/releng/issue/11531
* Fedora devel discussion: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/4S52TXRSMT4BMVZYKZ5GY3636ZTHPHAK/

**Comment by Conan-Kudo** [comment]
cc: @davide125, @michel-slm, @nirk

**Comment by Conan-Kudo** [comment]
cc: @chrismurphy

**Comment by Conan-Kudo** [comment]
cc: @cmurf

**Comment by Conan-Kudo** [comment]
For what it's worth, the difference of subvolumes instead of directories is just `btrfs subvolume (create|delete) </path/to/chroot/basedir>` instead of `(mkdir|rm -rf) </path/to/chroot/basedir>`.

**Comment by xsuchy** [comment]
Note: Mock already deletes all subvolumes it creates in chroot. See https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=1205564

**Comment by Conan-Kudo** [comment]
Right, so it just doesn't create them for the chroot path `/var/lib/mock/<rootname>`

**Comment by Conan-Kudo** [comment]
So would this mean making a btrfs_root plugin counterpart to the lvm_root plugin? https://github.com/rpm-software-management/mock/blob/main/mock/py/mockbuild/plugins/lvm_root.py

**Comment by xsuchy** [comment]
Yes.

**Comment by praiskup** [comment]
There was a typo, and @nirik originally wasn't pinged here.

I'm curious what is the motivation for keeping all the historical buildroots and not doing cleanups.  Isn't this too storage demanding, or is there some copy-on-write mechanism?

Also, how important this feature is for Koji administrators?  Seems like this is nothing new actually, would you actually use the new plugin (if it existed)?

**Comment by nirik** [comment]
buildroots are cleaned up/kept by kojid. It has some configuration options around that for how long to keep successfull ones around and how long to keep failed build ones around.

So, if mock made the buildroots btfs subvolumes, koji would have to learn about that in order to clean them up. 

So, ideally mock would have an option to do subvolumes and kojid would know how to clean them up based on it's settings. If this existed, I think fedora at least would be happy to use it.
